technical-docs/2021-05-remote-restore-redesign.md

```diff
@@ -0,0 +1,110 @@
+# Medusa Operator - Restore redesign document
+
+## Background
+
+Medusa operator doesn’t yet support remote restores: restoring a backup to another cluster.
+As we implement this feature, we would like to have it support multiple use cases that operators could expect to have working:
+
+- Restoring a medusa-operator based backup to another Cassandra install in the same k8s cluster
+![](img/medusa_k8s-new-restore-design-1.jpg)
+
+- Restoring a medusa-operator based backup to another Cassandra install in a different k8s cluster
+![](img/medusa_k8s-new-restore-design-2.jpg)
+
+- Restoring a Medusa backup (non k8s) to a k8s Cassandra install
+![](img/medusa_k8s-new-restore-design-3.jpg)
+
+
+
+
+## Sync CassandraBackup resources with storage
+The target solution will need to be able to sync the local CassandraBackup resources with the configured storage backend.
+The sync should allow:
+
+- Creating local missing resources from existing backups
+- Deleting local resources for missing backups
+- Deleting remote backups if a local resource was marked as deleted
+
+A single control loop should be able to handle all 3 operations.
+The sync should be required explicitly, meaning that a new CRD should be created to request synchronization to the medusa-operator, and that CRD should contain a prefix so that a cluster can list backups stored for another cluster in a multi-tenant storage bucket.
+
+The CassandraBackup resource should be updated, adding a delete field which would allow marking backups for deletion.
+
+The sync operation would:
+
+- List the backups in the storage bucket, through a gRPC call
+- Compare that list to the existing CassandraBackup resources
+- Create a CassandraBackup object for each missing backup
+- Delete local resources that don’t exist in the list of backups
+- Delete remote backups through gRPC for CassandraBackup objects marked for deletion, and upon success delete the resource.
+
+## CassandraBackup CRD changes
+CassandraBackup resources shouldn’t store the CassandraDatacenter spec anymore but rather the topology object generated by Medusa.  
```

Comment:
Can you provide an example of what the topology object looks like?

Reply:
Sure thing, it currently looks like this:
```
{"ip-179-31-16-38.us-west-2.compute.internal": {"tokens": [-1051417974279770378, -112095142485141927, ..., 9217379683620055011, 939597040143840721], "is_up": true}, 
 "ip-179-31-23-54.us-west-2.compute.internal": {"tokens": [-1080182033668782299, -1156652990105327691, ..., 9134535641516355167, 9216775765860996507], "is_up": true}, 
 "ip-179-31-30-139.us-west-2.compute.internal": {"tokens": [-1018393144648419592, -1342532530495729638, ..., 8846603596658721956, 8849342512812637165], "is_up": true}}
```

It should be augmented with the dc and rack information.

Reply:
Added the sample to the design doc.